### PR TITLE
Add analytical 2d cost function for pose graph (not used yet).

### PR DESCRIPTION
**Pull request by pifon2a**
Benchmark for costs after 2*10^7 evaluations with randomly pregenerated poses.

analytical 1713.24 ms
autodiff 3390.28 ms

### CODE REVIEW

**Comment by pifon2a** [comment]
Average over 10 pose graph optimization launches (Deutsches Museum 1. floor was loaded as an unfrozen state):             
                                            analytical             autodiff
jacobian computation:         0.324915             0.4290803333  // analytical jacobian worked ~32% faster.
total time                             2.954363667       3.247346333 
fraction of time in jac comp 0.1100644721     0.1322957539

File: cartographer/mapping/internal/optimization/cost_functions/spa_cost_function_2d.cc [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        common::NormalizeAngleDifference(
+            observed_relative_pose_.rotation().angle() - (end[2] - start[2]));
+    if (jacobians == NULL) return true;
+
+    const double weighted_cos_start_rotation =
+        translation_weight_ * cos_start_rotation;
+    const double weighted_sin_start_rotation =
+        translation_weight_ * sin_start_rotation;
+
+    if (jacobians[0] != NULL) {
```
Comment (gaschler) [inline cartographer/mapping/internal/optimization/cost_functions/spa_cost_function_2d.cc]: add comment why these are ordered like this

File: cartographer/mapping/internal/optimization/cost_functions/spa_cost_function_2d.cc [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      jacobians[0][0] = weighted_cos_start_rotation;
```
Comment (gaschler) [inline cartographer/mapping/internal/optimization/cost_functions/spa_cost_function_2d.cc]: I think the sign is wrong, in line 81 it's -sin so here it should be -cos.

**Comment by pifon2a** [comment]
PTAL

**Comment by ojura** [comment]
@pifon2a link is not public, is that intentional?

**Comment by pifon2a** [comment]
@ojura yes, it was just a rendered version of Sphinx documentation, to make the review easier.

**Pull request merged**